privacyidea/lib/framework.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+#
+#  2018-11-15   Friedrich Weber <friedrich.weber@netknights.it>
+#               Add a framework module to reduce the coupling to flask
+#
+# This code is free software; you can redistribute it and/or
+# modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+# License as published by the Free Software Foundation; either
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+from flask import current_app, g
+# We import the gettext function here and export it as ``_``.
+from flask_babel import gettext as _
+
+
+def get_application_local_store():
+    """
+    Get a dictionary which is local to the current flask application,
+    but shared among all threads.
+    :return: a Python dict
+    """
+    # We can use ``setdefault`` here because starting from
+    # Python 2.7.3 and 3.2.3, it is guaranteed to be atomic.
+    return current_app.config.setdefault('_app_local_store', {})
+
+
+def get_request_local_store():
+    """
+    Get a dictionary which is local to the current request. Thus, it
+    is not shared among threads.
+    :return: a Python dict
+    """
+    if '_request_local_store' not in g:
+        g._request_local_store = {}
+    return g._request_local_store
+
+
+def get_app_config_values():
```

Comment:
I was confused by ``get_app_config_values``  and ``get_app_config``.
I was expecting to get the complete config object with ``get_app_config``. But it was the other way round.

Should we put it into one function? Or should we do

get_app_config -> return the complete dictionary.
get_app_configvalue -> return only one value.

A function ``get_app_config`` would always return a dictionary. Otherwise we would not need the abstraction layer.

Could we drop ``get_app_configvalue`` and instead do:

~~~~python
value = get_app_config().get(key, None)
~~~~

Reply:
I agree, I also found that confusing :-)

``get_app_config`` already existed prior to this PR, I just moved it from ``privacyidea.lib.config`` to ``privacyidea.lib.framework``.

I think it would be nice to keep a dedicated function to retrieve *one* config value, because this is needed very often. The complete dictionary is only needed at one place, by the HSM initialization code.

So what do you think about just renaming

* ``get_app_config`` to ``get_app_config_value`` and
* ``get_app_config_values`` to ``get_app_config``

?

Reply:
Sounds good to me, if ``get_app_config`` returns the complete config and ``get_app_config_value`` returns a single value from the config.

Reply:
Alright, will do that!

Reply:
done in 6bf4167